docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,557 @@
+
+## Implementation
+
+### Storage
+
+We suggest to use the same storage mechanism proposed in approach A.
+
+### Peer Selection
+
+To achieve bucket-powered peer selection, we propose to create a standalone cache like component to manage buckets, such that it tracks peers by their IDs (_`peer.ID`_) and puts them in the correct bucket every time a `Sort` method is called on the component. The component, thus, maps each peer ID to a list of timestamps, each timestamp marking when did the bucketing component see this peer.
+
+The length of the list of timestamps will be equal to the amount of times the peer was seen, and the dates from oldest to newest will help the component sort the peer into appropriate buckets.
+
+Example:
+A list of 1 timestamp means it's been seen once, and if the timestamp was 3 hours ago, then the peer would be sorted into the "< 1 day ago" bucket.
+
+Depending on how many peers the node will be connected to, this could be the best peer it can get, or the worst.
+
+A better example of what a very good peer would be is:
+A peer that has a list of timestamps whose length is equal to 10, and the oldest timestamp was a week ago and the newest from an hour ago. This will make the peer fall into "< 1 week ago" bucket but also all the other recent buckets.
+
+We suggest to also turn the "amount of times seen" into buckets, such that we sort peers into buckets of "how many times" they were seen, modifying the selection criteria for good peers to become:
+
+* Peers that fall in most recent buckets + Peers that fall in highest seen count buckets
+
+**The interface**: We will go with the name `Bucketer` for the lack of a better one for the moment
+
+```go
+type Bucketer interface {
+    AddPeer(peer.ID, time.Time)
+    AddPeers([]peer.ID, time.Time)
+    Sort()
+    BestPeers() []peer.ID
+    PurgePeer(peer.ID)
+    PurgeAll()
+}
+```
+
+The implementer of this interface should rely on golang maps as the native abstraction for buckets, example:
+
+```go
+type bucketer struct {
+    timeBuckets map[BucketDuration][]peer.ID
+    seenBuckets map[int][]peer.ID
+    peersSeen map[peer.ID][]time.Time
+}
+```
+
+with `BucketDuration` being a new type to enumerate all possible time buckets, example:
+
+```go
+type BucketDuration string
+const (
+    LTOneHour = "< 1 hour"
+    LTOneDay = "< 1 day"
+    ...
+)
+```
+
+The `AddPeer` or `AddPeers`, and the `Sort` methods are intended to be called from within a procedure `fetchSortPersistPeers` that is called within a routine that periodically fetches peers from the peer store of the host (_`peerstore.Peerstore`_) which contains all the peers we are connected to.
+
+`fetchSortPersistPeers` will also be called on node shutdown.
+
+`fetchSortPersistPeers` will be as follows:
+
+```go
+func fetchSortPersistPeers(bcktr Bucketer, store PeerStore, host host.Host) {
```

Comment:
This function does too much. The routine that periodically fetches peers could maybe add to the bucket, even if its in another subroutine - and pass the peers to this function as a slice. This function can be called `persistPeers` and be responsible for sorting and persisting the peers to the store. Why is the storage not a method on the peerstore?

Also, maybe, add a method to the bucketer that returns the sorted best peers?

Reply:
In terms of storage, we have to distinguish between `Peerstore` (_libp2p's in-mem peerstore_) and the `PeerStore`, the badger store we define for this ADR's purposes.

If you read through the proposed code, the bucketer has a `BestPeers` method that returns the sorted best peers.

I am not entirely opposed to turning this method into `persistPeers([]peer.ID, host.Host)` such that it only receives best peers as a subroutine. This was intended to be an illustrative example and not necessarily a spec for how to write the function.